### PR TITLE
PriorityValue and PriorityLevel Memory Usage Optimization

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Implemented optimization to only construct a dictionary for priority levels if there is more than one in use. This drops memory consumption in most applications by a reasonable amount since most don't use multiple bindings on single properties (like animation and a local binding).

### CODE REVIEW

File: src/Avalonia.Base/Utilities/SingleOrDictionary.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    /// Stores either a single key value pair or constructs a dictionary when more than one value is stored.
+    /// </summary>
+    /// <typeparam name="TKey">The type of the key.</typeparam>
+    /// <typeparam name="TValue">The type of the value.</typeparam>
+    public class SingleOrDictionary<TKey, TValue> : IEnumerable<KeyValuePair<TKey, TValue>>
+    {
+        private KeyValuePair<TKey, TValue>? singleValue;
+        private Dictionary<TKey, TValue> dictionary;
+
+        private bool useDictionary = false;
```
Comment (grokys) [inline src/Avalonia.Base/Utilities/SingleOrDictionary.cs]: Is `useDictionary` needed? Isn't it the same as `dictionary != null`?

File: src/Avalonia.Base/Utilities/SingleOrDictionary.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+namespace Avalonia.Utilities
+{
+    /// <summary>
+    /// Stores either a single key value pair or constructs a dictionary when more than one value is stored.
+    /// </summary>
+    /// <typeparam name="TKey">The type of the key.</typeparam>
+    /// <typeparam name="TValue">The type of the value.</typeparam>
+    public class SingleOrDictionary<TKey, TValue> : IEnumerable<KeyValuePair<TKey, TValue>>
+    {
+        private KeyValuePair<TKey, TValue>? singleValue;
```
Comment (grokys) [inline src/Avalonia.Base/Utilities/SingleOrDictionary.cs]: Fields should begin with underscores.

**Comment by grokys** [comment]
This is a clever optimization! @wieslawsoltes @danwalmsley I'd be interested to see how much memory it saves in your apps.

**Comment by grokys** [comment]
An even cleverer optimization might be to add an `IPriorityValue` interface and have two implementations: one that contains just a single binding and the current one. But that would be a deeper change :D

**Comment by jkoritzinsky** [comment]
I decided to do the optimization this way so we can generalize it later if we see similar memory performance characteristics with other dictionaries. I can also generalize this to collections and not just dictionaries. My goal was to only change `PriorityValue` to create a `SingleOrDictionary` and not make any functional code changes.

**Comment by grokys** [comment]
Yes, it makes sense to make the change small and self-contained. We can always revisit later if we decide it's worth it - was just putting it out there ;)

**Comment by jkoritzinsky** [comment]
I agree 👍

**Comment by grokys** [comment]
By the way, do you have any figures as to the amount of memory saved?

**Comment by jkoritzinsky** [comment]
On ControlCatalog.Desktop immediately after it initially shows onscreen, almost a 200kB difference. And the difference will only get larger the larger the application.

I don't have a chance to check out my oxyplot demo again, but the difference there is probably a little under 1MB.

**Pull request merged**